clouddrift/wavelet.py

```diff
@@ -0,0 +1,801 @@
+    >>> z = np.random.random(1024) + 1j*np.random.random(1024)
+    >>> wtz_p, wtz_n = morse_wavelet_transform(z, 3, 4, np.array([2*np.pi*0.2]), complex=True)
+
+    The same result as above can be otained by applying the Morse transform on the real and imaginary
+    component of z and recombining the results as follows for the "bandpass" normalization:
+    >>> wtz_real = morse_wavelet_transform(np.real(z)), 3, 4, np.array([2*np.pi*0.2]))
+    >>> wtz_imag = morse_wavelet_transform(np.imag(z)), 3, 4, np.array([2*np.pi*0.2]))
+    >>> wtz_p, wtz_n = (wtz_real + 1j*wtz_imag) / 2, (wtz_real - 1j*wtz_imag) / 2
+
+    For the "energy" normalization, the analytic and anti-analytic components are obtained as follows
+    with this alternative method:
+    >>> wtz_real = morse_wavelet_transform(np.real(z)), 3, 4, np.array([2*np.pi*0.2]))
+    >>> wtz_imag = morse_wavelet_transform(np.imag(z)), 3, 4, np.array([2*np.pi*0.2]))
+    >>> wtz_p, wtz_n = (wtz_real + 1j*wtz_imag) / np.sqrt(2), (wtz_real - 1j*wtz_imag) / np.sqrt(2)
+
+    The input signal can have an arbitrary number of dimensions but its ``time_axis`` must be
+    specified if it is not the last:
+
+    >>> x = np.random.random((1024,10,15))
+    >>> wtx = morse_wavelet_transform(x, 3, 4, np.array([2*np.pi*0.2]), time_axis=0)
+
+    The default way to handle the boundary conditions is to mirror the ends points
+    but this can be changed by specifying the chosen boundary method:
+
+    >>> x = np.random.random((10,15,1024))
+    >>> wtx = morse_wavelet_transform(x, 3, 4, np.array([2*np.pi*0.2]), boundary="periodic")
+
+    This function can be used to complete a time-frequency analysis of the input signal by specifying
+    a range of randian frequencies using the ``morse_logspace_freq`` function as an example:
+
+    >>> x = np.random.random(1024)
+    >>> gamma = 3
+    >>> beta = 4
+    >>> radian_frequency = morse_logspace_freq(gamma, beta, np.shape(x)[0])
+    >>> wtx = morse_wavelet_transform(x, gamma, beta, radian_frequency)
+
+    Raises
+    ------
+    ValueError
+        If the time axis is outside of the valid range ([-1, np.ndim(x)-1]).
+        If boundary optional argument is not in ["mirror", "zeros", "periodic"]``.
+        If normalization optional argument is not in ["bandpass", "energy"]``.
+
+    See Also
+    --------
+    :func:`morse_wavelet`, `wavelet_transform`, `morse_logspace_freq`
+
+    """
+    # time_axis must be in valid range
+    if time_axis < -1 or time_axis > len(x.shape) - 1:
+        raise ValueError(
+            f"time_axis ({time_axis}) is outside of the valid range ([-1,"
+            f" {len(x.shape) - 1}])."
+        )
+    # generate the wavelet
+    wavelet, _ = morse_wavelet(
+        np.shape(x)[time_axis],
+        gamma,
+        beta,
+        radian_frequency,
+        normalization=normalization,
+        order=order,
+    )
+
+    # apply the wavelet transform, distinguish complex and real cases
+    if complex:  # np.all(np.isreal(x))
```

Comment:
Why is there the `# np.all(np.isreal(x))` comment here?

Reply:
Also, is there cases where we would want to pass a `complex` signal but use `complex=False` here?

Reply:
ok removed the comment which was a previous implementation. Yes, the savy/knowledgeable user might want to do that, and this is why it is not automatically tested but passed as an argument.